src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts

```diff
@@ -0,0 +1,963 @@
+      offset,
+      rowsPerImage,
+      bytesPerRow,
+      changeBeforePass
+    );
+    const appliedCopySize = this.appliedCopySize(width, height, depth, changeBeforePass);
+
+    switch (method) {
+      case 'WriteTexture': {
+        this.device.defaultQueue.writeTexture(
+          appliedCopyView,
+          partialData,
+          appliedDataLayout,
+          appliedCopySize
+        );
+
+        break;
+      }
+      case 'CopyB2T': {
+        const buffer = this.device.createBuffer({
+          mappedAtCreation: true,
+          size: align(partialData.byteLength, 4),
+          usage: GPUBufferUsage.COPY_SRC,
+        });
+        new Uint8Array(buffer.getMappedRange()).set(partialData);
+        buffer.unmap();
+
+        const encoder = this.device.createCommandEncoder();
+        encoder.copyBufferToTexture(
+          { buffer, ...appliedDataLayout },
+          appliedCopyView,
+          appliedCopySize
+        );
+        this.device.defaultQueue.submit([encoder.finish()]);
+
+        break;
+      }
+      default:
+        unreachable();
+    }
+  }
+
+  // We check an appropriate part of the texture against the given data.
+  // Used directly with PartialCopyT2B check method (for a subpart of the texture)
+  // and by fullCheck function with FullCopyT2B check method (for the whole texture).
+  checkData(
+    { texture, mipLevel, origin }: TextureCopyViewWithRequiredOrigin,
+    textureDataLayout: GPUTextureDataLayout,
+    format: SizedTextureFormat,
+    checkSize: GPUExtent3DDict,
+    expected: Uint8Array,
+    changeBeforePass: ChangeBeforePass = 'none'
+  ): void {
+    const buffer = this.device.createBuffer({
+      size: align(expected.byteLength, 4), // this is necessary because we need to copy and map data from this buffer
+      usage: GPUBufferUsage.COPY_SRC | GPUBufferUsage.COPY_DST,
+    });
+
+    const { offset, rowsPerImage, bytesPerRow } = textureDataLayout;
+    const { x, y, z } = origin;
+    const { width, height, depth } = checkSize;
+
+    const appliedCopyView = this.appliedCopyView(texture, x, y, z, mipLevel, changeBeforePass);
+    const appliedDataLayout = this.appliedDataLayout(
+      offset,
+      rowsPerImage,
+      bytesPerRow,
+      changeBeforePass
+    );
+    const appliedCheckSize = this.appliedCopySize(width, height, depth, changeBeforePass);
+
+    const encoder = this.device.createCommandEncoder();
+    encoder.copyTextureToBuffer(
+      appliedCopyView,
+      { buffer, ...appliedDataLayout },
+      appliedCheckSize
+    );
+    this.device.defaultQueue.submit([encoder.finish()]);
+
+    for (const texel of this.iterateBlockRows(checkSize, origin, format)) {
+      const rowOffset = this.getTexelOffsetInBytes(textureDataLayout, format, texel, origin);
+      const rowLength = this.bytesInACompleteRow(checkSize.width, format);
+      this.expectContents(buffer, expected.slice(rowOffset, rowOffset + rowLength), rowOffset);
```

Comment:
We should expect something about the data in between rows and in between images, too, although I don't know what it is. (Probably that it's zeroed? Maybe that it's preserved?)

Reply:
I added validation that the other values are preserved.